### ISSUE TITLE
Allow for custom docx page settings in markdown via raw code segment

### ISSUE THREAD

**Issue by mjfs** [jgm/pandoc#7255]
`Pandoc` always appends default `sectPr` segment to the end of `document.xml` in `docx` output file format. A user can change this for the document as a whole by proving custom template via `--reference-doc` switch.

When dealing with different page orientations (e.g. table output in landscape), it would be useful to have control over the segment appended by default.

One such use case is adding a table to the end of the `docx` document under appendixes in landscape while the rest of the document still having default portrait orientation.

Bellow is an example of what is proposed to allow user to do (e.g. via special command line switch disabling segment inserted by default or via advanced `docx` templating support) in `markdown` document, either directly or with `Lua` script and custom syntax. Enable using this segment as the last thing in the `markdown` document and not cause blank pages or errors:

	```{=openxml}
	<w:sectPr>
		<w:pgSz w:w="11906" w:h="16838" w:code="9"/>
		<w:pgMar w:top="2268" w:right="1276" w:bottom="2268" w:left="1418" w:header="567" w:footer="510" w:gutter="0"/>
		<w:cols w:space="708"/>
		<w:docGrid w:linePitch="360"/>
	</w:sectPr>
	```

**Comment by jgm** [comment]
I'm a bit confused.  Currently pandoc inserts either an empty `<w:sectPr/>` element or the one defined in the reference.docx (if there is one).

So, can't you just use a reference.docx that defines the sectPr you want?

It looks as if we could simply omit the empty `<w:sectPr />` element when none is specified in the reference.docx; I don't think it's necessary.  So maybe that would be enough for your needs?

**Comment by jgm** [comment]
Notes:
See http://officeopenxml.com/WPsection.php
This sectPr that appears last in the body affects the whole document.
A sectPr that affects a section of the document, not the document as a whole, apparently goes inside the pPr element for the last paragraph in the section:
``` xml
<w:p>
<w:pPr>
<w:sectPr>
<w:pgSz w:w="12240" w:h="15840"/>
</w:sectPr>
</w:pPr>
</w:p>
```
So that's the structure you'd need to have sectPr affect just part of a document. There's no way to achieve that currently.

**Comment by mjfs** [comment]
Yes, what you are describing in the two comments is exactly the issue, I should probably state it more explicitly - in `docx` one can have many segments and rules pertaining to a section of the document (e.g. borders, orientation, margins etc. for specific chapter). Ability to skip the default section/document settings (and enter it manually via raw segment) would enable more control.

One can now already enter other section declarations via raw element  in `Pandoc` for `docx` output, but being unable to skip/override the default results in two consecutive declarations when trying to manage the last section (and consequently usually an additional blank page or some similar oddity).

**Comment by jgm** [comment]
It would be easy to remove the empty sectPr that is added by default. Would that do the trick for you, or is more necessary?

**Comment by mjfs** [comment]
Not just skipping default empty `sectPr`, also skipping the one in the template while still allowing other customization provided via `--reference-doc` (at least styles).

The bullet proof option would be to check if `sectPr`exists (i.e. after markdown processing) and if not, add it. For my use case however not adding one at all (neither empty one nor the one from template) would completely work.

**Comment by ilcpm** [comment]
@mjfs 我们（非pandoc官方）想尝试做一个专门为md转换docx设计的pandoc插件，你的这个问题我们在我们的项目中用一些方法解决了。不过毕竟只是两个非计科的本科生，所以项目进度比较慢，而且出现了比较严重的项目管理问题，目前仓库比较混乱（有几个私有仓库），初稿应该是6月底或者7月会整理好后对外发布，可以关注一下https://github.com/ilcpm/pandoc_markdown2docx_instuction

**Comment by ilcpm** [comment]
In our project, we find a way to deal with this problem.

for pandoc ,it will copy the first `w:sectPr` tag from "reference.docx" to the end of the output docx file.

so the problem is how to write our own `w:sectPr` tag in the input.md.

to deal with this, we defined several `\newSection{}` command in markdown file and deal it in the filter, here is an example:

if you want to use 3 different `w:sectPr`

- define a generic command called `\newSection{}`
- put different page information into `{}` and use `\newcommand` (provided by pandoc) to rename the new commands to other names for short
- and finally just deal with the `\newSection{}` command in your filter

---

this is how we do it

write like this at the beginning of the markdown file

```
\newcommand\newSectionAbstract{\newSection{<w:pgSz w:w="11907" w:h="16840" w:code="9"/><w:pgMar w:top="1701" w:right="1418" w:bottom="1418" w:left="1418" w:header="907" w:footer="851" w:gutter="567"/>}}
```

this will transfer the page size and page margin information to the command `\newSection{}` and get a short command named `\newSectionAbstract{}` (we use this `w:sectPr` after our abstract)

and in our filter, add the prefix and suffix to the page information to deal with the `\newSection{}` command, so it return a `RawBlock` like

```
<w:p><w:pPr><w:sectPr>
<w:pgSz w:w="11907" w:h="16840" w:code="9"/><w:pgMar w:top="1701" w:right="1418" w:bottom="1418" w:left="1418" w:header="907" w:footer="851" w:gutter="567"/>
</w:sectPr></w:pPr></w:p>
```

so, finally you can write `\newSectionAbstract{}` to get a `w:sectPr` with your custom page size and page margin information, if you want use 3 different `w:sectPr`, just use `\newcommand` 3 times to get 3 different `w:sectPr`

---

this method may be a little bothering, but surely it works, and we even set different footers and headers for the output document.

**Comment by jgm** [comment]
I love it: defining LaTeX macros that return Word XML!  Twisted.  Your filter must also change the format on these RawBlocks from "latex" to "openxml", right?

**Comment by ilcpm** [comment]
> I love it: defining LaTeX macros that return Word XML! Twisted. Your filter must also change the format on these RawBlocks from "latex" to "openxml", right?

yeah, that was think up by my colleague. (I want to set a config file for our filter before that.)

we defined many tex command in our markdown file and transfer then into "openxml", such as the "tab"(w:tab), "newpage", and even the fieldcode (many functions still have problems, we are struggling with them)

we defined 3 "section separator" in our "head.md", they use the same page size and page margin, but was set to different footers and headers. And even customized the title of the "toc"

```
\newcommand\pageInfo{<w:pgSz w:w="11907" w:h="16840" w:code="9"/><w:pgMar w:top="1701" w:right="1418" w:bottom="1418" w:left="1418" w:header="907" w:footer="851" w:gutter="567"/><w:footnotePr><w:numFmt w:val="decimalEnclosedCircleChinese"/><w:numRestart w:val="eachPage"/></w:footnotePr>}

\newcommand\newSectionAbstract{\newSection{<w:pgNumType w:fmt="upperRoman" w:start="1"/><w:headerReference w:type="even" r:id="rId9"/><w:headerReference w:type="default" r:id="rId12"/><w:footerReference w:type="even" r:id="rId16"/><w:footerReference w:type="default" r:id="rId15"/><w:headerReference w:type="first" r:id="rId10"/><w:footerReference w:type="first" r:id="rId14"/>\pageInfo}}

\newcommand\newSectionTOC{\newSection{<w:pgNumType w:fmt="upperRoman" /><w:headerReference w:type="even" r:id="rId13"/><w:headerReference w:type="default" r:id="rId11"/>\pageInfo}}

\newcommand\toc{\tocRaw{目    录}}

```

(the third separator is inherited from the reference.docx)

sincerely thanks to pandoc and your works :)

**Comment by mjfs** [comment]
@jgm: I am using the same approach as the one that @ilcpm mentioned. I noticed `pagebreak` filter a couple of years back in `lua-filters` repository and later modified that to `\page{setting}` filter that covers other page related settings (e.g. page/section breaks, orientation, margins). It is similar to what @ilcpm is describing (only with shorter, more case specific syntax).

Actually, repository that @ilcpm linked above also contains `pagebreak` filter, so this might even be the spark behind the idea in both of our cases. I left the latex related part out of the open issue since it is not relevant to the root cause and a minimum working example with raw segment seemed clearer.

Unfortunately, `lua` filter and/or a switch passed to Pandoc can not currently solve the open issue: page settings in the `docx` template will be appended to the output `docx` and can not be overridden.

One can currently either adjust the the template to fit the last page/segment or, as mentioned, deal with oddity at the end of the document (i.e. blank page or unwanted page/margin settings) when customizing document segments via inline instructions. I have used a simple external script to delete default segment in the past, but this is cumbersome, less portable and adds additional security questions.

**Comment by jfrank-nih** [comment]
@mjfs, do you have a copy of that forked `lua-filters` repository with the `\page{setting}`? I don't see it on your profile. I've been working on a similar thing to implement a `\sectionbreak` that can also be a `sectionbreak{landscape}`. 

I hit the same problem with pandoc inserting a `sectPr` at the end that is overriding the `sectPr` I appended to the end of the document.
